### ISSUE TITLE
question re: Trickster expected performance

### ISSUE THREAD

**Issue by TimSimmons** [Comcast/trickster#281]
:wave: hello! I finally had a second to try out Trickster and I wanted to give a snapshot of my experience and see if it lined up with your expectations. I pulled the latest beta
```
$ ./trickster-1.0.9.linux-amd64 --version
1.0.9
```
and loaded up an instance with the following config
```
[main]
  instance_id = 0
...

[origins]
  [origins.default]
....
  [origins.general]
    type = "prometheus"
    scheme = "http"
    host = "<prometheus url running in the same datacenter>"
    path_prefix = "/general"
    api_path = "/api/v1/"
    ignore_no_cache_header = true
    timeseries_retention_factor = 10000
    timeseries_eviction_method = "oldest"
    fast_forward_disable = false
    backfill_tolerance_secs = 0
    timeout_secs = 180
    keep_alive_timeout_secs = 300
    max_idle_conns = 20
    cache_name = "default"
    is_default = false

[caches]
  [caches.default]
    type = "memory"
    compression = true
    timeseries_ttl_secs = 21600
    object_ttl_secs = 30
    fastforward_ttl_secs = 15
    [caches.default.index]
      reap_interval_secs = 120
      flush_interval_secs = 5
      max_size_bytes = 536870912
      max_size_backoff_bytes = 16777216
      max_size_objects = 0
      max_size_backoff_objects = 100
```

I then made a bigish query directly to the Prometheus instance
```
$ time curl '<prometheus>/general/api/v1/query_range?query=up&start=1569272293&end=1569358696&step=1m' > data
real	0m7.937s
```
It took ~8 seconds and returned ~160 MB of data

Then I made the same query to trickster:
```
$ time curl 'localhost:9091/general/api/v1/query_range?query=up&start=1569272293&end=1569358696&step=1m' > data
real	1m27.102s
```
The query took significantly longer, although I do see in the trickster logs that the upstream query took ~the same time from Prometheus.
```
time=2019-09-24T21:00:12.837143096Z app=trickster caller=proxy/engines/access_logs.go:23 level=debug event="upstream request" size=169740515 durationMS=10185 originName=general handlerName=QueryRangeHandler method=GET userAgent=curl/7.35.0 code=200 originType=prometheus uri="http://...:80/general/api/v1/query_range?end=1569358680&query=up&start=1569272280&step=1m"
```

I then made the query again against trickster, running locally I figured the response would be near instant now that the results were cached. 

```
$ time curl 'localhost:9091/general/api/v1/query_range?query=up&start=1569272293&end=1569358696&step=1m' > data

real	0m51.747s
```

but the query still took a pretty long time. I can see in the logs that no upstream request was made:
```
time=2019-09-24T21:03:00.180012987Z app=trickster caller=cache/memory/memory.go:63 level=debug event="memorycache cache retrieve" cacheKey=pandora.internal.digitalocean.com:80.d028ab238acdbb6a1b9cd8fc32e1db16.sz
time=2019-09-24T21:03:00.180138106Z app=trickster caller=proxy/engines/cache.go:41 level=debug event="decompressing cached data" cacheKey=pandora.internal.digitalocean.com:80.d028ab238acdbb6a1b9cd8fc32e1db16.sz
time=2019-09-24T21:03:50.892128366Z app=trickster caller=proxy/engines/deltaproxycache.go:323 level=debug event="delta routine completed" cacheKey=pandora.internal.digitalocean.com:80.d028ab238acdbb6a1b9cd8fc32e1db16 cacheStatus=hit reqStart=1569272280 reqEnd=1569358680
```

I can see the cache being used, about ~35 MB of data after the initial query, and trickster using up to ~2.5 GB of memory while I was using it, I set up another instance of Prometheus to scrape the instance while I was testing:
![tricksterperf](https://i.imgur.com/Kc821si.png)

I saw a similar result on a much smaller scale, changing my step from `1m` to `1h` I saw the initial trickster query take ~2.5s, and then consistently ~1s. The query over the network to the actual Prometheus instance also took ~1s.

I also performed a similar test with `compression = false`, and saw similar results, except that the cache usage was ~160 MB, as expected.

So my observations were as follows:

1. Trickster did cache my large/small query results, but it took a significant amount of time (>10x the time of the actual query) on the original request to cache the data *before* returning the query to me.
2. Once cached, requests to an instance running on localhost were still significantly slower than re-performing the request to the origin Prometheus instance over the network.
3. Trickster memory usage was significant, much higher than the cache itself (~10x). Even accounting for the recent Go changes that make `process_resident_memory_bytes` it was a lot. and `htop` reported ~1.5 GB of usage.

I'm curious if these results match your expectations based on my rather rudimentary test, or if something in the neolithic way I chose to test might be rendering my results invalid. In my infrastructure 100+ MB is definitely a large query, but not unheard of, it's one of the things I hoped I could utilize Trickster to mitigate, as those queries often hurt my instances quite a bit.

Thanks!

**Comment by jranson** [comment]
@TimSimmons thanks for this report. I'm on business travel and haven't had a chance to look at this, but should get to it over the next couple of days. Sorry for the delay. This is definitely not the behavior or performance we'd hope to see, so we'll get to the bottom of it.

**Comment by jranson** [comment]
@TimSimmons, i took a look at this and am not seeing this behavior (caveat: I am testing everything locally on my mac via the darwin binary, and via localhost rather than an actual multi-host routed network; however network locale should not affect the trickster memory internals when processing data). 

Your example query is exactly 24h3s in range size. Generally speaking, querying for 1 day's worth of `?query=up` at a 60s resolution, as in your example, should return about 25K of data (1440 timestamps of approx 17 bytes each) nearly instantaneously; so 160MB seems quite excessive, especially when querying prom directly -- that payload size should represent about 17y worth of data. That said, I ran a test query for 1y worth of data @ 60s resolution (using Trickster's `promsim` package as the upstream origin in order to generate data for that far back on the fly), and did not encounter any issues. In that case the dataset was 9MB and returned in under 4s via trickster.  Running the `up` query for the last 17y against promsim via trickster took about 1m11s on the initial miss, and 1m10s on the second run (partial hit with the vast majority of the data being uncached). 

What I will do in the next 24h is to give this a go via the linux binary on a linux host, rather than darwin on a mac, just to make sure it's not a linux-specific issue.

Questions: 

Can you describe your deployment environment a little more (e.g., is this on Amazon EC2 or some other cloud computing platform, on baremetal, via kube, docker, etc.. 

Secondly what flavor and version of linux are you running?

And lastly, do you have any kind of prom aggregation solution upstream of Trickster (thanos, promxy, m3, etc) that could potentially be placing a lot more data into to the response than expected? I am not sure why else your payload might be so large and would definitely like to understand that. If you don't mind posting some of the payload of your 160MB output (esp. any "metric" sections of the output that would help understand how many different time series are returned), that would be great; that way we can understand what is going on with the data size, separately from the data processing latency and memory utilization observations.

Thanks again for your report and helping to get to the bottom of this!

**Comment by jranson** [comment]
Ah - i see that the `up` query will return one series per target, so that 25K would be per series. Doing the math, it looks like you might have about 6500 targets in prom, does that sound right?  I am able to reproduce this with promsim with those conditions (`up` @ 6500 targets, 1d, 1m resolution) and will take a look today. Thanks for this!

**Comment by TimSimmons** [comment]
Yup `6910` timeseries as of right now! Thanks for looking into it. The reason I chose this as a sample is I routinely have issues with folks querying lots of timeseries like this, especially with histogram metrics, so I wanted to see how it'd work caching large result sets like this. Excited to see what you find out :eyes:

**Comment by jranson** [comment]
I think the basic issue is that all of trickster's data merging is being done by iterating the series in a serial fashion; I will get that bit moved into a goroutine and see how that works out. We should have beta 10 / GA RC1 cut next week, and that will include the optimization.

**Comment by TimSimmons** [comment]
That makes sense, in my case at least I'm ok with Trickster using a ton of CPU/RAM as it'd be a completely separate component from everything else. To answer your question from earlier the topology on this particular request was
DigitalOcean Droplet -> HW Kubernetes Pod (nginx proxy) -> DigitalOcean Droplet, all in the same datacenter.

**Comment by jranson** [comment]
Hey Tim!  Sorry for taking longer than expected on this one. We have released 1.0 Beta 10 which should fully resolve this particular bottleneck (as far as I was able to test and confirm per your instructions). If you wouldn't mind giving it a try, we'd love to know how things are working out for you. And be sure to take a look at all the new goodies that are in this release too! Thanks again for the report and being so patient while we completed all of the other unrelated features to get this release ready. If you are still seeing the issue, definitely re-open this issue and we'll revisit.
